### PR TITLE
Added typeORM migration howto.

### PR DESCRIPTION
**Pull request by ambroiseRabier**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/nestjs/docs.nestjs.com/blob/master/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[x] Other... Please describe: documentation
```

## What is the current behavior?

Issue Number: https://github.com/nestjs/typeorm/issues/123 and 
https://github.com/nestjs/typeorm/issues/33

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

Resolve https://github.com/nestjs/typeorm/issues/123
Resolve https://github.com/nestjs/typeorm/issues/33

I have made a working example here: https://github.com/ambroiseRabier/typeorm-nestjs-migration-example

### CODE REVIEW

**Comment by ambroiseRabier** [comment]
Is there any interest in content of the pull request? (if no, it is not worth spending my time trying to resolve CI issues)

**Comment by kamilmysliwiec** [comment]
@ambroiseRabier I'll review your PR as soon as possible. Sorry for the delay!

**Comment by webcat12345** [comment]
This is really helpful

File: content/techniques/typeorm-migration.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+```
+
+#### Usage
+1. `npm run typeorm:migrate <myEntity-migration>`
+2. Check your migration queries in `src/migrations`.
+3. `npm run start:dev` or `npm run start:prod` or `npm run typeorm:run`
+
+If everything went well, you have up to date entites and a `migrations` table listing applied migrations.
+
+#### Additionnal information
```
Comment (gbirke) [inline content/techniques/typeorm-migration.md]:
```suggestion
#### Additional information
```

**Comment by prateekkathal** [comment]
@johnbiundo I suggest closing this, since this is irrelevant and a setting already present in the TypeORM module. 

A better way to add this to the documentation would be a direct link to the [TypeORM documentation about Migrations](https://github.com/typeorm/typeorm/blob/master/docs/migrations.md).

I really don't feel a new page in the documentation is required for this. The PR is super old & has failing pipelines.

**Comment by johnbiundo** [comment]
@prateekkathal Thanks for the feedback.  I'll try to find some time to study it further.

**Comment by ambroiseRabier** [comment]
I do not understand why it is failing the pipeline:

```
Error: ENOENT: no such file or directory, open '/opt/build/repo/src/app/homepage/pages/techniques/typeorm-migration/typeorm-migration.component.html'
8:00:22 PM:     at Object.fs.openSync (fs.js:646:18)
```

Is not very helpful log. Anyway I can see there is mixed feelings about if that should be into NestJS documentation, feel free to close it.

**Comment by prateekkathal** [comment]
> I do not understand why it is failing the pipeline:
> 
> ```
> Error: ENOENT: no such file or directory, open '/opt/build/repo/src/app/homepage/pages/techniques/typeorm-migration/typeorm-migration.component.html'
> 8:00:22 PM:     at Object.fs.openSync (fs.js:646:18)
> ```
> 
> Is not very helpful log. Anyway I can see there is mixed feelings about if that should be into NestJS documentation, feel free to close it.

The error is basically related to not having a component bound that markdown file which would generate an HTML file. You will have to create a new component file and then bind it with a selector and then later add that component to the `TechniquesModule`.

@ambroiseRabier We could merge this in if you pull in from the latest branch & maybe make the suggested changes. 

I leave the rest on @johnbiundo

**Comment by ambroiseRabier** [comment]
Seem like the CI is passing, I moved the doc from techniques to recipes though, not quite sure where it should go.

**Comment by Nabellaleen** [comment]
1h lost looking for such a doc !

It is very useful, so please review and merge this PR ! :D

File: content/recipes/typeorm-migration.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+**src/app.module.ts**
+```typescript
+import { DynamicModule, Module } from '@nestjs/common';
+import { AppController } from './app.controller';
+import { AppService } from './app.service';
+import { TypeOrmModule } from '@nestjs/typeorm';
+import * as ormconfig from './ormconfig';
+
+export function DatabaseOrmModule(): DynamicModule {
```
Comment (jmcdo29) [inline content/recipes/typeorm-migration.md:19]: Would this be better as a function or a `const` like
```ts
export const DatabaseModule: DynamicModule = TypeOrmModule.forRoot(ormConfig);
```
This way it stays in the same vein that _usually_ raw functions are not used for the `imports` array.

File: content/recipes/typeorm-migration.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  // We are using migrations, synchronize should be set to false.
+  synchronize: false,
+
+  // Run migrations automatically,
+  // you can disable this if you prefer running migration manually.
+  migrationsRun: true,
+  logging: true,
+  logger: 'file',
+
+  // Allow both start:prod and start:dev to use migrations
```
Comment (jmcdo29) [inline content/recipes/typeorm-migration.md:71]: As `start:dev` is now an alias for `nest start --watch` which runs `tsc --watch` under the hood, should this be updated to mention that only `dist + '/migrations/**/*.js'` need to be looked for, or should it be kept like this for the case of people still using `ts-node`?

**Comment by johnbiundo** [comment]
@jmcdo29 Sorry I've been somewhat MIA this week.  I haven't had a chance to really read through it, but I trust others' judgement on its usefulness.  However, at a quick glance, one thing is that it is not consistent with our doc style.  I think it would be useful to reformat it in our more standard approach.

I think, at this point, I'll ask @kamilmysliwiec  if he agrees with the idea of adding this page. If he does, I think we can find a way to do the formatting.  I'd be happy to at the very least do some editing, if not write some of the prose portion.

**Comment by kamilmysliwiec** [comment]
I don't think we should provide an additional Recipe for this because we already have a dedicated section in the Database chapter:

image

If we want to improve it, we can consider adding more information there instead. However, I'm not really sure what would be useful for the community so I'm open to suggestions.

**Comment by ambroiseRabier** [comment]
Feel free to add this in a more consistent way to NestJS Database documentation chapter, I do not think I be able to correctly follow the writing style. Giving @jmcdo29, I might be a little out of date with my recipe, so you might want to test it with latest version of NestJS.

**Comment by jmcdo29** [comment]
I know I've seen questions around migrations in the discord. What we may want to do is mention explicitly that with using migrations and the TypeORM CLI, you cannot use the `forRootAsync()` method without duplicating configurations. I know it's already mentioned that you cannot use the DI services, but this would drive home the point that they're two separate systems entirely.

**Comment by kamilmysliwiec** [comment]
> What we may want to do is mention explicitly that with using migrations and the TypeORM CLI, you cannot use the forRootAsync() method without duplicating configurations. 

This might be confusing. In fact, you can use the `forRootAsync()` and don't duplicate configurations by leveraging configuration namespaces https://docs.nestjs.com/techniques/configuration#configuration-namespaces :D

**Comment by raphaelbadia** [comment]
Hey, I'm trying to setup migrations and I'm a bit lost : it seems that I need to setup a typeormconfig.json file to make the migration work but if I do so, the docs warn me that webpack won't work and that I cannot leverage `autoLoadEntities` and `retryDelay`. A PR like this would help a lot !

**Comment by ronen-agranat** [comment]
If you’re reading this, chances are that you are wanting to set up database configurations for NestJS, that also work for your TypeORM migrations.

I wrote a blog post on how you can do this: https://www.ronenagranat.com/2021-03-01/nestjs-typeorm-environment-config-migration.html

With thanks to the author of this PR and all contributors on this discussion.

Please let me know if you have any feedback or if it's helpful!

**Pull request closed without merging**